includes/traits/llms-trait-sales-page.php

```diff
@@ -0,0 +1,140 @@
+<?php
+/**
+ * LifterLMS Sales Page trait
+ *
+ * @package LifterLMS/Traits
+ *
+ * @since [version]
+ * @version [version]
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * LifterLMS Sales Page trait.
+ *
+ * **This trait should only be used by classes that extend from the {@see LLMS_Post_Model} class.**
+ * **Classes that use this trait must call {@see LLMS_Trait_Sales_Page::construct_sales_page} in their constructor.**
+ *
+ * @since [version]
+ *
+ * @property int    $sales_page_content_page_id WP Post ID of the WP page to redirect to when $sales_page_content_type is 'page'.
+ * @property string $sales_page_content_type    Sales page behavior [none,content,page,url].
+ * @property string $sales_page_content_url     Redirect URL for a sales page, when $sales_page_content_type is 'url'.
+ */
+trait LLMS_Trait_Sales_Page {
+	/**
+	 * @inheritdoc
+	 */
+	abstract protected function add_properties( $props = array() );
```

Comment:
Thank you for catching that `LLMS_Trait_Audio_Video_Embed` needs to have abstract `add_properties()` and `get()` methods. Having an [abstract trait method](https://www.php.net/manual/en/language.oop5.traits.php#language.oop5.traits.abstract) imposes a requirement on the exhibiting class to implement the method, else a fatal error is triggered. It's also good documentation.

Unfortunately, properties can not be declared as abstract and can not be [redefined](https://www.php.net/manual/en/language.oop5.traits.php#language.oop5.traits.properties) without triggering a fatal error. I did not find a simple way to make PHP enforce the requirement. Options:
1. Use `property_exists()` and then `trigger_error()` if `$model_post_type` does not exist.
2. Add notes in the DocBlock and hope for the best.

@eri-trabiccolo and @thomasplevy, what are your thoughts on properties that the trait depends on?

Reply:
>  Having an abstract trait method imposes a requirement on the exhibiting class to implement the method, else a fatal error is triggered.


Yeah sure, otherwise a fatal error will be triggered as well when trying to execute a non existing method. Right?
I mean maybe you're right and it's better to have that fatal triggering asap and not depending on whether that method will be ever executed?!
I see it's better for documentation in any case.
 

> Unfortunately, properties can not be declared as abstract and can not be redefined without triggering a fatal error. I did not find a simple way to make PHP enforce the requirement. Options:
> 
> Use property_exists() and then trigger_error() if $model_post_type does not exist.
> Add notes in the DocBlock and hope for the best.


Not sure, also because thinking more about it, in that case we don't only need that `$model_post_type` exists, but that it's also defined (`isset`) and actually even not empty.
Let's go with DocBlock only. IMHO.